demos/howtos/myula.py

```diff
@@ -0,0 +1,217 @@
+#
+# To illustrate MYULA, we will consider :math:`g(x) = \texttt{strength_reg} \  TV(x) = \texttt{strength_reg} \|\nabla x \|_{2, 1}`,
+# where :math:`\texttt{strength_reg}` is the regularization parameter which
+# controls the regularization strength induced by TV.
+
+# %%
+# Bayesian model definition
+# -------------------------
+# Then consider the following Bayesian model
+#
+# .. math::
+#    \begin{align*}
+#    \mathbf{x} &\sim \exp (- \texttt{strength_reg} \|\nabla x \|_{2,1})\\
+#    \mathbf{y} &\sim \mathcal{N}(\mathbf{A}\mathbf{x}, \texttt{sigma2}\,\mathbf{I}) \ ,
+#    \end{align*}
+#
+# with :math:`\texttt{sigma2}=0.05^2`.
+
+# %%
+# Likelihood definition
+# ---------------------
+# We first specify the data distribution as follows:
+sigma2=0.05**2
+y=cuqi.distribution.Gaussian(A, sigma2)
+# %%
+# Then we can define the likelihood with
+likelihood=y(y=y_obs)
+
+# %%
+# Implicit prior definition
+# ---------------------------------
+# To apply MYULA, we need to define the implicit prior :math:`\pi_{\texttt{stength_smooth}}(x)`.
+# Evaluating this surrogate prior is doable but too intensive from
+# a computational point of view as it requires to solve an optimization problem.
+# However to apply MYULA, we only require access to
+# :math:`\operatorname{prox}_{\texttt{strength_reg}\ TV}^{\texttt{strength_smooth}}`.
+#
+# As suggested by Durmus et al. (https://arxiv.org/pdf/1612.07471), we set the
+# smoothing parameter :math:`\texttt{strength_smooth} \approx \texttt{sigma2}`,
+# ie :math:`\texttt{strength_smooth}= 0.5 \ \texttt{sigma2}`.
+#
+# We set the regularization parameter to :math:`\texttt{stength_reg}=10`.
+
+# %%
+# Regularization and smoothing parameters definition
+strength_reg=10
+strength_smooth=0.5*sigma2
+# %%
+# To estimate :math:`\operatorname{prox}_{\texttt{strength_reg}\  TV}^{\texttt{strength_smooth}}`
+# we use the implementation provided by Scikit-Image. But we can use any solver
+# to compute this quantity.
+#
+# We emphasize that we have for any :math:`g`
+#
+# .. math::
+#       \operatorname{prox}_{\texttt{strength_reg}\  g}^{\texttt{strength_smooth}} = \operatorname{prox}_{g}^{\texttt{weight}} ,
+#
+# with :math:`\texttt{weight} = \texttt{strength_reg} \times  \texttt{strength_smooth}`.
+from skimage.restoration import denoise_tv_chambolle
+def prox_g(x, strength_reg=None, strength_smooth=None):
+    weight=strength_reg*strength_smooth
+    return denoise_tv_chambolle(x, weight=weight, max_num_iter=100), None
+# %%
+# We save all the important variables into the variable
+# :math:`\texttt{denoiser_kwargs}`.
+denoiser_kwargs={}
+denoiser_kwargs["strength_reg"]=strength_reg
+denoiser_kwargs["strength_smooth"]=strength_smooth
+# %%
+# Now we can define our implicit prior.
+denoise_regularizer = DenoiseRegularizer(
+    prox_g,
+    strength_smooth=strength_smooth,
+    denoiser_kwargs=denoiser_kwargs,
+    geometry=likelihood.model.domain_geometry
+)
+
```

Comment:
It will helpful to somehow illustrate what this denoiseregularizer "does", before using it in the more involved posterior sampling setting.  Can you add, at this point in the demo, a small demonstration of calling denoise_regularizer.regularize() on a noisy image and seeing its denoising effect by comparing the noisy and denoised image?

Reply:
That is a really good idea. I will take care of that.
But maybe "regularize()" is a bit missleading because we do not apply the regularizer itself but the denoiser. We recently changed from "denoise()" to "regularize()" but we are not 100% convinced. What do you think?

Reply:
Please can you add the demonstration with the current proposed syntax. Once we settle on naming etc., we can adjust it.

Reply:
Done! The demo is updated
(Good point)

Reply:
Thank you both for the discussion.
After the redesign, this discussion mostly refers to the old design, and the demo is updated to illustrate the restoration operation as @remilaumont pointed out